### PR TITLE
Avoid vector copies in range-for loops in opt/types.cpp

### PR DESCRIPTION
**Pull request by dneto0**
Also be more explicit about iterated types in other range-for loops.

This subsumes #1260

### CODE REVIEW

File: source/opt/types.cpp [hunk 1]
```diff
@@ -173,7 +173,7 @@ bool Type::operator==(const Type& other) const {
 
 void Type::GetHashWords(std::vector<uint32_t>* words) const {
   words->push_back(kind_);
-  for (auto d : decorations_) {
+  for (const auto& d : decorations_) {
```
Comment (dneto0) [inline source/opt/types.cpp]: Use reference, not copy of vector.

File: source/opt/types.cpp [hunk 2]
```diff
@@ -455,12 +455,12 @@ std::string Struct::str() const {
...
 void Struct::GetExtraHashWords(std::vector<uint32_t>* words) const {
-  for (auto t : element_types_) {
+  for (auto* t : element_types_) {
     t->GetHashWords(words);
   }
-  for (auto pair : element_decorations_) {
+  for (const auto& pair : element_decorations_) {
     words->push_back(pair.first);
-    for (auto d : pair.second) {
+    for (const auto& d : pair.second) {
```
Comment (dneto0) [inline source/opt/types.cpp]: Use a reference, not an expensive copy.

File: source/opt/types.cpp [hunk 3]
```diff
@@ -455,12 +455,12 @@ std::string Struct::str() const {
 }
 
 void Struct::GetExtraHashWords(std::vector<uint32_t>* words) const {
-  for (auto t : element_types_) {
+  for (auto* t : element_types_) {
```
Comment (atgoo) [inline source/opt/types.cpp]: auto and auto* don't have much difference, except for more type safety.

File: source/opt/types.cpp [hunk 4]
```diff
@@ -407,7 +407,7 @@ void RuntimeArray::GetExtraHashWords(std::vector<uint32_t>* words) const {
 
 Struct::Struct(const std::vector<Type*>& types)
     : Type(kStruct), element_types_(types) {
-  for (auto* t : types) {
+  for (const auto* t : types) {
```
Comment (atgoo) [inline source/opt/types.cpp]: This looks like an empty loop for NDEBUG.

Comment (dneto0) [reply source/opt/types.cpp]: Yes.

File: source/opt/types.cpp (hunk 3 above)
Comment (dneto0) [reply source/opt/types.cpp]: Yes.  After seeing #1260 I went through and annotated things more explicitly.  Having the * there makes it clearer this is an inexpensive iteration.   So it helps verifiability.

**Comment by dneto0** [comment]
Rebased.  Keeping approval.

**Comment by alan-baker** [comment]
Rebased and submitted as 3ef4bb600f2ac09fda73671617e7793b5891afea.

Tested manually on windows.

**Pull request closed without merging**